import pytest

from gitfourchette.forms.commitdialog import CommitDialog
from gitfourchette.forms.identitydialog import IdentityDialog
from gitfourchette.forms.signatureform import SignatureOverride
from gitfourchette.graphview.commitlogmodel import CommitLogModel
from gitfourchette.nav import NavLocator
from . import reposcenario
from .util import *


def testCommit(tempDir, mainWindow):
    wd = unpackRepo(tempDir)
    writeFile(F"{wd}/a/a1.txt", "a1\nPENDING CHANGE\n")  # unstaged change
    rw = mainWindow.openRepo(wd)

    qlvClickNthRow(rw.dirtyFiles, 0)
    QTest.keyPress(rw.dirtyFiles, Qt.Key.Key_Return)
    assert qlvGetRowData(rw.dirtyFiles) == []
    assert qlvGetRowData(rw.stagedFiles) == ["a/a1.txt"]
    rw.commitButton.click()

    dialog: CommitDialog = findQDialog(rw, "commit")
    QTest.keyClicks(dialog.ui.summaryEditor, "Some New Commit")

    dialog.ui.revealSignature.click()

    enteredDate = QDateTime.fromString("1999-12-31 23:59:00", "yyyy-MM-dd HH:mm:ss")
    sigUI = dialog.ui.signature.ui
    qcbSetIndex(sigUI.replaceComboBox, "author")
    sigUI.nameEdit.setText("Custom Author")
    sigUI.emailEdit.setText("custom.author@example.com")
    sigUI.timeEdit.setDateTime(enteredDate)

    dialog.accept()

    headCommit = rw.repo.head_commit
    assert headCommit.message == "Some New Commit"
    assert headCommit.author.name == "Custom Author"
    assert headCommit.author.email == "custom.author@example.com"
    assert headCommit.author.time == enteredDate.toSecsSinceEpoch()
    assert headCommit.committer.name == TEST_SIGNATURE.name

    assert len(headCommit.parents) == 1
    diff = rw.repo.diff(headCommit.parents[0], headCommit)
    patches: list[Patch] = list(diff)
    assert len(patches) == 1
    assert patches[0].delta.new_file.path == "a/a1.txt"


def testCommitUntrackedFileInEmptyRepo(tempDir, mainWindow):
    wd = unpackRepo(tempDir, "TestEmptyRepository")
    touchFile(F"{wd}/SomeNewFile.txt")
    rw = mainWindow.openRepo(wd)

    qlvClickNthRow(rw.dirtyFiles, 0)
    QTest.keyPress(rw.dirtyFiles, Qt.Key.Key_Return)

    assert qlvGetRowData(rw.dirtyFiles) == []
    assert qlvGetRowData(rw.stagedFiles) == ["SomeNewFile.txt"]

    rw.commitButton.click()
    dialog: CommitDialog = findQDialog(rw, "commit")
    QTest.keyClicks(dialog.ui.summaryEditor, "Initial commit")
    dialog.accept()

    rows = qlvGetRowData(rw.graphView, CommitLogModel.CommitRole)
    commit: Commit = rows[-1].peel(Commit)
    assert commit.message == "Initial commit"


def testCommitMessageDraftSavedOnCancel(tempDir, mainWindow):
    wd = unpackRepo(tempDir)
    reposcenario.stagedNewEmptyFile(wd)
    rw = mainWindow.openRepo(wd)

    rw.commitButton.click()
    dialog: CommitDialog = findQDialog(rw, "commit")
    assert dialog.ui.summaryEditor.text() == ""
    assert dialog.getOverriddenSignatureKind() == SignatureOverride.Nothing
    QTest.keyClicks(dialog.ui.summaryEditor, "hoping to save this message")
    dialog.reject()
    assert rw.state.uiPrefs.draftCommitMessage == "hoping to save this message"
    assert rw.state.uiPrefs.draftCommitSignatureOverride == SignatureOverride.Nothing

    rw.commitButton.click()
    dialog: CommitDialog = findQDialog(rw, "commit")
    assert dialog.ui.summaryEditor.text() == "hoping to save this message"
    dialog.ui.revealSignature.click()
    dialog.ui.signature.ui.replaceComboBox.setCurrentIndex(2)
    dialog.reject()
    assert rw.state.uiPrefs.draftCommitMessage == "hoping to save this message"
    assert rw.state.uiPrefs.draftCommitSignatureOverride == SignatureOverride.Both
    assert rw.state.uiPrefs.draftCommitSignature is not None

    rw.commitButton.click()
    dialog: CommitDialog = findQDialog(rw, "commit")
    assert dialog.ui.summaryEditor.text() == "hoping to save this message"
    assert dialog.getOverriddenSignatureKind() == SignatureOverride.Both
    dialog.accept()  # Go through with the commit this time

    # Ensure nothing remains of the draft after a successful commit
    rw.commitButton.click()
    acceptQMessageBox(rw, "empty commit")
    dialog: CommitDialog = findQDialog(rw, "commit")
    assert dialog.ui.summaryEditor.text() == ""
    assert dialog.getOverriddenSignatureKind() == SignatureOverride.Nothing
    dialog.reject()


def testAmendCommit(qtbot, tempDir, mainWindow):
    oldMessage = "Delete c/c2-2.txt"
    newMessage = "amended commit message"
    newAuthorName = "Jean-Michel Tartempion"
    newAuthorEmail = "jmtartempion@example.com"

    wd = unpackRepo(tempDir)
    reposcenario.stagedNewEmptyFile(wd)
    rw = mainWindow.openRepo(wd)

    oldHeadCommit = rw.repo.head_commit

    # Kick off amend dialog
    rw.amendButton.click()

    dialog: CommitDialog = findQDialog(rw, "amend")
    assert dialog.ui.summaryEditor.text() == oldMessage
    dialog.ui.summaryEditor.setText(newMessage)
    dialog.ui.revealSignature.setChecked(True)
    dialog.ui.signature.ui.nameEdit.setText(newAuthorName)
    dialog.ui.signature.ui.emailEdit.setText(newAuthorEmail)
    with qtbot.waitSignal(dialog.destroyed):  # upon exiting context, wait for dialog to be gone
        dialog.accept()

    headCommit = rw.repo.head_commit
    assert headCommit.id != oldHeadCommit.id
    assert headCommit.message == newMessage
    assert headCommit.author.name == newAuthorName
    assert headCommit.author.email == newAuthorEmail
    assert headCommit.committer.name == TEST_SIGNATURE.name
    assert headCommit.committer.email == TEST_SIGNATURE.email

    # Ensure no error dialog boxes after operation
    assert not mainWindow.findChildren(QDialog)

    assert not rw.graphView.currentCommitId  # "uncommitted changes" should still be selected
    assert rw.stagedFiles.isVisibleTo(rw)
    assert rw.dirtyFiles.isVisibleTo(rw)
    assert not rw.committedFiles.isVisibleTo(rw)


def testAmendCommitDontBreakRefresh(qtbot, tempDir, mainWindow):
    wd = unpackRepo(tempDir)
    reposcenario.stagedNewEmptyFile(wd)
    rw = mainWindow.openRepo(wd)

    # Kick off amend dialog
    rw.jump(NavLocator.inWorkdir())
    triggerMenuAction(rw.graphView.makeContextMenu(), "amend")

    # Amend HEAD commit without any changes, i.e. just change the timestamp.
    dialog: CommitDialog = findQDialog(rw, "amend")
    with qtbot.waitSignal(dialog.destroyed):  # upon exiting context, wait for dialog to be gone
        dialog.accept()

    # Ensure no errors dialog boxes after operation (e.g. "commit not found")
    assert not mainWindow.findChildren(QDialog)


def testEmptyCommitRaisesWarning(tempDir, mainWindow):
    wd = unpackRepo(tempDir)
    rw = mainWindow.openRepo(wd)
    rw.commitButton.click()
    rejectQMessageBox(rw, "create.+empty commit")


def testCommitWithoutUserIdentity(tempDir, mainWindow):
    clearSessionwideIdentity()

    wd = unpackRepo(tempDir)
    rw = mainWindow.openRepo(wd)

    assert "user.name" not in rw.repo.config
    assert "user.email" not in rw.repo.config

    rw.commitButton.click()
    acceptQMessageBox(rw, "create.+empty commit")

    identityDialog = findQDialog(rw, "identity")
    assert isinstance(identityDialog, IdentityDialog)
    identityOK = identityDialog.ui.buttonBox.button(QDialogButtonBox.StandardButton.Ok)
    assert not identityOK.isEnabled()
    identityDialog.ui.nameEdit.setText("Archibald Haddock")
    identityDialog.ui.emailEdit.setText("1e15sabords@example.com")
    assert identityOK.isEnabled()
    identityDialog.accept()

    commitDialog = findQDialog(rw, "commit")
    assert isinstance(commitDialog, CommitDialog)
    commitDialog.ui.summaryEditor.setText("ca geht's mol?")
    commitDialog.accept()

    headCommit = rw.repo.head_commit
    assert headCommit.message == "ca geht's mol?"
    assert headCommit.author.name == "Archibald Haddock"
    assert headCommit.author.email == "1e15sabords@example.com"


def testCommitStableDate(tempDir, mainWindow):
    wd = unpackRepo(tempDir)
    writeFile(F"{wd}/a/a1.txt", "a1\nPENDING CHANGE\n")  # unstaged change
    rw = mainWindow.openRepo(wd)

    rw.commitButton.click()
    acceptQMessageBox(rw, "empty commit")

    dialog: CommitDialog = findQDialog(rw, "commit")
    dialog.ui.summaryEditor.setText("hold on a sec...")

    # Wait for next second before confirming.
    # Commit time should not depend on when the dialog is accepted.
    QTest.qWait(1001)
    dialog.accept()

    headCommit = rw.repo.head_commit
    assert headCommit.message == "hold on a sec..."
    assert signatures_equalish(headCommit.author, headCommit.committer)


def testAmendAltersCommitterDate(tempDir, mainWindow):
    wd = unpackRepo(tempDir)
    writeFile(F"{wd}/a/a1.txt", "a1\nPENDING CHANGE\n")  # unstaged change
    rw = mainWindow.openRepo(wd)

    headCommit = rw.repo.head_commit
    rw.amendButton.click()

    dialog: CommitDialog = findQDialog(rw, "amend")
    dialog.ui.summaryEditor.setText("hold on a sec...")
    dialog.accept()

    amendedHeadCommit = rw.repo.head_commit
    assert amendedHeadCommit.message == "hold on a sec..."
    assert signatures_equalish(amendedHeadCommit.author, headCommit.author)
    assert not signatures_equalish(amendedHeadCommit.committer, headCommit.committer)
    assert not signatures_equalish(amendedHeadCommit.author, amendedHeadCommit.committer)
    assert amendedHeadCommit.author.name != TEST_SIGNATURE.name
    assert amendedHeadCommit.committer.name == TEST_SIGNATURE.name
    assert amendedHeadCommit.committer.time > amendedHeadCommit.author.time


def testCommitDialogJumpsToWorkdir(tempDir, mainWindow):
    wd = unpackRepo(tempDir)
    reposcenario.fileWithStagedAndUnstagedChanges(wd)
    rw = mainWindow.openRepo(wd)

    oid1 = Oid(hex="0966a434eb1a025db6b71485ab63a3bfbea520b6")
    rw.jump(NavLocator.inCommit(oid1))

    triggerMenuAction(mainWindow.menuBar(), r"repo/commit")
    findQDialog(rw, r"commit").reject()
    assert NavLocator.inUnstaged("a/a1.txt").isSimilarEnoughTo(rw.navLocator)

    rw.jump(NavLocator.inStaged("a/a1.txt"))
    triggerMenuAction(mainWindow.menuBar(), r"repo/commit")
    findQDialog(rw, r"commit").reject()
    assert NavLocator.inStaged("a/a1.txt").isSimilarEnoughTo(rw.navLocator)


@pytest.mark.parametrize("method", ["graphkey", "graphcm"])
def testCheckoutCommitDetachedHead(tempDir, mainWindow, method):
    wd = unpackRepo(tempDir)
    rw = mainWindow.openRepo(wd)
    repo = rw.repo

    for oid in [Oid(hex="0966a434eb1a025db6b71485ab63a3bfbea520b6"),
                Oid(hex="6db9c2ebf75590eef973081736730a9ea169a0c4"),
                ]:
        rw.jump(NavLocator.inCommit(oid))

        if method == "graphcm":
            triggerMenuAction(rw.graphView.makeContextMenu(), r"check.?out")
        elif method == "graphkey":
            QTest.keySequence(rw.graphView, "Return")
        else:
            raise NotImplementedError(f"unknown method {method}")

        dlg = findQDialog(rw, "check.?out commit")
        dlg.findChild(QRadioButton, "detachedHeadRadioButton", Qt.FindChildOption.FindChildrenRecursively).setChecked(True)
        dlg.accept()

        assert repo.head_is_detached
        assert repo.head_commit_id == oid

        assert rw.graphView.currentCommitId == oid, "graphview's selected commit has jumped around"


def testCommitOnDetachedHead(tempDir, mainWindow):
    wd = unpackRepo(tempDir)

    oid = Oid(hex='1203b03dc816ccbb67773f28b3c19318654b0bc8')

    with RepoContext(wd) as repo:
        repo.checkout_commit(oid)

    rw = mainWindow.openRepo(wd)

    assert rw.repo.head_is_detached
    assert rw.repo.head.target == oid

    displayedCommits = qlvGetRowData(rw.graphView, Qt.ItemDataRole.UserRole)
    assert rw.repo.head_commit in displayedCommits

    rw.commitButton.click()
    acceptQMessageBox(rw, "create.+empty commit")
    commitDialog: CommitDialog = findQDialog(rw, "commit")
    commitDialog.ui.summaryEditor.setText("les chenilles et les chevaux")
    commitDialog.accept()

    assert rw.repo.head_is_detached
    assert rw.repo.head.target != oid  # detached HEAD should no longer point to initial commit

    newHeadCommit = rw.repo.head_commit
    assert newHeadCommit.message == "les chenilles et les chevaux"

    displayedCommits = qlvGetRowData(rw.graphView, Qt.ItemDataRole.UserRole)
    assert newHeadCommit in displayedCommits


def testRevertCommit(tempDir, mainWindow):
    wd = unpackRepo(tempDir)
    rw = mainWindow.openRepo(wd)

    oid = Oid(hex="c9ed7bf12c73de26422b7c5a44d74cfce5a8993b")
    rw.jump(NavLocator.inCommit(oid))
    triggerMenuAction(rw.graphView.makeContextMenu(), "revert")

    rw.jump(NavLocator.inWorkdir())
    assert qlvGetRowData(rw.stagedFiles) == ["c/c2-2.txt"]
    assert rw.repo.status() == {"c/c2-2.txt": FileStatus.INDEX_NEW}


def testCherrypick(tempDir, mainWindow):
    wd = unpackRepo(tempDir)

    with RepoContext(wd) as repo:
        repo.checkout_local_branch("no-parent")

    oid = Oid(hex='ac7e7e44c1885efb472ad54a78327d66bfc4ecef')  # "First a/a1"

    rw = mainWindow.openRepo(wd)

    rw.jump(NavLocator.inCommit(oid))
    triggerMenuAction(rw.graphView.makeContextMenu(), "cherry")

    assert rw.diffArea.fileStackPage() == "workdir"
    assert rw.repo.status() == {"a/a1.txt": FileStatus.INDEX_NEW}

    acceptQMessageBox(rw, "cherry.+success.+commit")

    dialog: CommitDialog = findQDialog(rw, "commit")
    assert dialog.ui.summaryEditor.text() == "First a/a1"
    assert dialog.getOverriddenSignatureKind() == SignatureOverride.Author

    dialog.accept()

    headCommit = rw.repo.head_commit
    assert headCommit.message == "First a/a1"
    rw.jump(NavLocator.inCommit(headCommit.id))
    assert qlvGetRowData(rw.committedFiles) == ["a/a1.txt"]


def testCherrypickDud(tempDir, mainWindow):
    wd = unpackRepo(tempDir)
    rw = mainWindow.openRepo(wd)

    oid = Oid(hex="f73b95671f326616d66b2afb3bdfcdbbce110b44")
    rw.jump(NavLocator.inCommit(oid))
    triggerMenuAction(rw.graphView.makeContextMenu(), "cherry")
    acceptQMessageBox(rw, "nothing to cherry.?pick.+already")


def testAbortCherrypick(tempDir, mainWindow):
    wd = unpackRepo(tempDir)

    with RepoContext(wd) as repo:
        repo.checkout_local_branch("no-parent")

    oid = Oid(hex='ac7e7e44c1885efb472ad54a78327d66bfc4ecef')  # "First a/a1"

    rw = mainWindow.openRepo(wd)

    rw.jump(NavLocator.inCommit(oid))
    triggerMenuAction(rw.graphView.makeContextMenu(), "cherry")
    assert rw.diffArea.fileStackPage() == "workdir"
    assert rw.repo.status() == {"a/a1.txt": FileStatus.INDEX_NEW}
    rejectQMessageBox(rw, "cherry.+success.+commit")

    assert rw.repo.state() == RepositoryState.CHERRYPICK
    assert "First a/a1" in rw.state.uiPrefs.draftCommitMessage
    assert rw.mergeBanner.isVisibleTo(rw)
    assert re.search(r"cherry.+conflicts fixed", rw.mergeBanner.label.text(), re.I | re.S)
    assert "abort" in rw.mergeBanner.button.text().lower()

    # Abort cherrypick
    rw.mergeBanner.button.click()
    acceptQMessageBox(rw, "abort.+cherry.+a/a1")

    assert rw.repo.state() == RepositoryState.NONE
    assert rw.repo.status() == {}
    assert rw.state.uiPrefs.draftCommitMessage == ""


def testNewTag(tempDir, mainWindow):
    newTag = "cool-tag"

    wd = unpackRepo(tempDir)
    rw = mainWindow.openRepo(wd)
    assert newTag not in rw.repo.listall_tags()

    oid = Oid(hex='ac7e7e44c1885efb472ad54a78327d66bfc4ecef')  # "First a/a1"

    rw.jump(NavLocator.inCommit(oid))
    triggerMenuAction(rw.graphView.makeContextMenu(), "tag this commit")

    dlg: QDialog = findQDialog(rw, "new tag")
    lineEdit = dlg.findChild(QLineEdit)
    QTest.keyClicks(lineEdit, newTag)
    dlg.accept()

    assert newTag in rw.repo.listall_tags()


def testDeleteTag(tempDir, mainWindow):
    tagToDelete = "annotated_tag"

    wd = unpackRepo(tempDir)
    rw = mainWindow.openRepo(wd)
    assert tagToDelete in rw.repo.listall_tags()

    node = rw.sidebar.findNodeByRef(f"refs/tags/{tagToDelete}")
    menu = rw.sidebar.makeNodeMenu(node)
    triggerMenuAction(menu, "delete")

    acceptQMessageBox(rw, "delete tag")
    assert tagToDelete not in rw.repo.listall_tags()
